packages/tangle/messagemetadata.go

```diff
@@ -126,6 +150,113 @@ func (m *MessageMetadata) SolidificationTime() time.Time {
 	return m.solidificationTime
 }
 
+// IsBooked returns true if the message represented by this metadata is booked. False otherwise.
+func (m *MessageMetadata) IsBooked() (result bool) {
+	m.bookedMutex.RLock()
+	defer m.bookedMutex.RUnlock()
+	result = m.booked
+
+	return
+}
+
+// IsEligible returns true if the message represented by this metadata is eligible. False otherwise.
+func (m *MessageMetadata) IsEligible() (result bool) {
+	m.eligibleMutex.RLock()
+	defer m.eligibleMutex.RUnlock()
+	result = m.eligible
+
+	return
+}
+
+// SetBooked sets the message associated with this metadata as booked.
+// It returns true if the booked status is modified. False otherwise.
+func (m *MessageMetadata) SetBooked(booked bool) (modified bool) {
+	m.bookedMutex.Lock()
+	defer m.bookedMutex.Unlock()
+
+	if m.booked == booked {
+		return false
+	}
+
+	m.booked = booked
+	m.SetModified()
+	modified = true
+
+	return
+}
+
+// TimestampOpinion returns the timestampOpinion of the given message metadata.
+func (m *MessageMetadata) TimestampOpinion() (timestampOpinion TimestampOpinion) {
+	m.timestampOpinionMutex.RLock()
+	defer m.timestampOpinionMutex.RUnlock()
+	return m.timestampOpinion
+}
+
+// SetTimestampOpinion sets the timestampOpinion flag.
+// It returns true if the timestampOpinion flag is modified. False otherwise.
+func (m *MessageMetadata) SetTimestampOpinion(timestampOpinion TimestampOpinion) (modified bool) {
+	m.timestampOpinionMutex.RLock()
+	if !m.timestampOpinion.Equal(timestampOpinion) {
+		m.timestampOpinionMutex.RUnlock()
+
+		m.timestampOpinionMutex.Lock()
+		if !m.timestampOpinion.Equal(timestampOpinion) {
+			m.timestampOpinion = timestampOpinion
+			m.SetModified()
+			modified = true
+		}
+		m.timestampOpinionMutex.Unlock()
+
+	} else {
+		m.timestampOpinionMutex.RUnlock()
+	}
```

Comment:
Double locking 🤢

Reply:
how exactly?

Reply:
I mean this rlock and lock business. It should just be something like 
```
func (m *MessageMetadata) SetBooked(booked bool) (modified bool) {
	m.bookedMutex.Lock()
	defer m.bookedMutex.Unlock()

	if m.booked == booked {
		return false
	}

	m.booked = booked
	m.SetModified()
	modified = true

	return
}```

Reads much easier :)

Reply:
true, but then this has to acquire the write lock even if the value is not modified. I can live with that :D

Reply:
Between line 200 and 202 another thread could have acquired the read lock also entered the condition and raced to win the write lock?

Reply:
It could, that's why there's another if check when the write lock is acquired in line 203 to avoid a race condition. Another reason why it's so ugly :D

Reply:
I see, there is another check. No bug.